erlcloud_ses: support (most of) the rest of the AWS SES API

nalundgaard opened the pull request:
Hi,

I work for [Alert Logic](https://github.com/alertlogic), and we use a [fork](https://github.com/alertlogic/erlcloud) of erlcloud for some of our work. I have been working with SES, and the need has arisen for us to leverage some SES functionality that isn't exposed by the `erlcloud_ses` module.

As a result, I have added most of the missing functions for the SES API to `erlcloud_ses`. The remaining methods I have not implemented were not needed by me, and I wasn't sure I wanted to mess with handling the url-encoded JSON mess with the SES `Policy` stuff. So, there are a few unimplemented methods that are deprecated by AWS, all of the Policies actions, and SendRawEmail.

These can easily be added in the future. I've opted not to implement these changes as `erlcloud_ses2` because the changes are generally backwards-compatible, and the only non-backwards-compatible thing that I know of (parsing the send_email response) is likely a change to a section of the result that I **hope** no one was using (a big naked XML string). More on that below.

I've endeavored to adhere to the style of the existing module and use patterns from elsewhere in erlcloud as much as possible. I thought I would submit this to you guys for your feedback. Thanks for looking!

## Commit notes

Methods not implemented:
  * DeleteIdentityPolicy
  * DeleteVerifiedEmailAddress (deprecated; use DeleteIdentity)
  * GetIdentityPolicies
  * ListIdentityPolicies
  * ListVerifiedEmailAddresses (deprecated; use ListIdentities)
  * PutIdentityPolicy
  * SendRawEmail
  * VerifyEmailAddress (deprecated; use VerifyEmailIdentity)

All other methods have been implemented. I have tested each method against the SES API using my own AWS account. This is backwards-compatible with the existing function (SendEmail was the only one), only I have added XML parsing to all methods—so, now, send_email returns `{ok, [{message_id, string()}]}` instead of a tuple of `ok` and a giant, naked XML body. If any client used this part of the response (e.g., some downstream parsing), this change will break that functionality.

Other functions in the API generally return proplists, with title case parameters from AWS converted to erlang atoms where possible. Other values are converted to strings, integers, floats, and booleans as appropriate for the corresponding values in the AWS response.

I've also added the ability to pass single items to lists. This means that, for example, in the `bcc_addresses` field of `To`, I can pass a single email address. The same is true for any function that expects a list of email addresses, identities, etc.

Finally, errors: now, `erlcloud_ses` returns parsed errors when AWS responds with XML. It also returns erlang errors when invalid parameters (that cannot be encoded) are passed to it. Here are some examples:

```erlang
erlcloud_ses:set_identity_feedback_forwarding_enabled("email@example.com", false).
{error,{erlcloud_ses,{invalid_parameter_value,"Feedback notification topic is not set."}}}
```
```erlang
erlcloud_ses:set_identity_dkim_enabled("example.com", trye).
** exception error: {erlcloud_ses,{invalid_parameter,{dkim_enabled,trye}}}
```

ransomr commented:
Thank you for the PR - may be a few days before I get a chance to take a detailed look.

nalundgaard commented:
@ransomr Cool! we'll be developing from a branch with these changes, so no rush on our end. I look forward to reading your feedback.

nalundgaard commented:
@ransomr I just remembered this afternoon that the eunit tests are in the `test` directory, and my change broke the existing one (due mainly to attempting to parse the XML that wasn't mecked by the test).

I fixed that test, and added coverage for all my added functions. I think the unit tests are broken right now due to a bug in `erlcloud_sns_tests` that I didn't spend any time looking into, but this works for me:

```sh
$ rebar eunit suites=erlcloud_ses_tests
==> erlcloud (eunit)
Compiled test/erlcloud_ses_tests.erl
  All 16 tests passed.
```

I should have included this in my PR to begin with, so it's handy that you didn't have time to review right away!

ransomr commented:
Thanks - glad you added test functions. As of my last commit all eunit tests were passing for me.

nalundgaard commented:
I get this with the current master, and of course this PR which doesn't affect this error.

```sh
$ make eunit
==> meck (compile)
==> jsx (compile)
==> lhttpc (compile)
==> erlcloud (compile)
==> erlcloud (eunit)
Compiled src/erlcloud_xml.erl
src/erlcloud_util.erl:38: Warning: crypto:md5/1 is deprecated and will be removed in in a future release; use crypto:hash/2
Compiled src/erlcloud_util.erl
Compiled src/erlcloud_sts.erl
Compiled src/erlcloud_ses.erl
Compiled src/erlcloud_sqs.erl
Compiled src/erlcloud_sns.erl
src/erlcloud_sdb.erl:33: Warning: export_all flag enabled - all functions will be exported
Compiled src/erlcloud_sdb.erl
Compiled src/erlcloud_retry.erl
Compiled src/erlcloud_mon.erl
Compiled src/erlcloud_kinesis_impl.erl
Compiled src/erlcloud_kinesis.erl
Compiled src/erlcloud_s3.erl
Compiled src/erlcloud_httpc.erl
Compiled src/erlcloud_iam.erl
Compiled src/erlcloud_http.erl
Compiled src/erlcloud_elb.erl
Compiled src/erlcloud_mturk.erl
Compiled src/erlcloud_ddb_util.erl
Compiled src/erlcloud_ddb_impl.erl
Compiled src/erlcloud_ddb1.erl
Compiled src/erlcloud_ddb.erl
Compiled src/erlcloud_ddb2.erl
Compiled src/erlcloud_cloudtrail.erl
Compiled src/erlcloud_as.erl
Compiled src/erlcloud.erl
Compiled src/erlcloud_aws.erl
Compiled test/erlcloud_ses_tests.erl
Compiled test/erlcloud_sns_tests.erl
Compiled test/erlcloud_s3_tests.erl
test/erlcloud_sdb_tests.erl:4: Warning: export_all flag enabled - all functions will be exported
Compiled test/erlcloud_sdb_tests.erl
Compiled src/erlcloud_ec2.erl
Compiled test/erlcloud_kinesis_tests.erl
test/erlcloud_iam_tests.erl:74: Warning: variable 'Key' exported from 'case' (line 67)
test/erlcloud_iam_tests.erl:78: Warning: variable 'Value' exported from 'case' (line 67)
Compiled test/erlcloud_iam_tests.erl
Compiled test/erlcloud_ddb_util_tests.erl
test/erlcloud_ec2_tests.erl:67: Warning: variable 'Key' exported from 'case' (line 60)
test/erlcloud_ec2_tests.erl:71: Warning: variable 'Value' exported from 'case' (line 60)
Compiled test/erlcloud_ec2_tests.erl
Compiled test/erlcloud_cloudtrail_tests.erl
Compiled test/erlcloud_ddb_tests.erl
Compiled test/erlcloud_aws_tests.erl
Compiled test/erlcloud_as_tests.erl
Compiled test/erlcloud_ddb2_tests.erl
undefined
*** instantiation of subtests failed ***
**in function jsx_decoder:value/4 (src/jsx_decoder.erl, line 234)
in call from erlcloud_aws:get_credentials_from_metadata/1 (src/erlcloud_aws.erl, line 262)
in call from erlcloud_aws:update_config/1 (src/erlcloud_aws.erl, line 211)
in call from erlcloud_aws:aws_request2/7 (src/erlcloud_aws.erl, line 75)
in call from erlcloud_aws:aws_request_xml2/7 (src/erlcloud_aws.erl, line 40)
in call from erlcloud_sns:sns_xml_request/3 (src/erlcloud_sns.erl, line 515)
in call from erlcloud_sns:publish/5 (src/erlcloud_sns.erl, line 353)
in call from erlcloud_sns_tests:defaults_to_http/1 (test/erlcloud_sns_tests.erl, line 27)
**error:badarg


undefined
*** context setup failed ***
**in function meck_proc:start/2 (src/meck_proc.erl, line 94)
  called as start(erlcloud_httpc,[])
in call from erlcloud_sns_tests:start/0 (test/erlcloud_sns_tests.erl, line 18)
**error:{already_started,<0.674.0>}
# tons more of  **error:{already_started,<0.674.0>}
```

This is with Erlang R16B03-1 and rebar:
```sh
rebar 2.5.1 R16B03-1 20141016_194954 git 2.5.1
```

ransomr commented:
Interesting. I've only tried on R17 recently, so I'm guessing that is the issue.

ransomr commented:
Thanks for the PR. Especially for adding tests and documentation.
